cas-server-core/src/main/java/org/jasig/cas/authentication/support/DefaultCasAttributeEncoder.java

```diff
@@ -0,0 +1,214 @@
+     *
+     * @param attributes the attributes
+     * @param service the service
+     */
+    protected void encodeAttributesInternal(final Map<String, Object> attributes, final RegisteredService service) {
+        try {
+            if (!initializeCipherBasedOnServicePublicKey(service)) {
+                return;
+            }
+            encodeAndEncryptCredentialPassword(attributes);
+            encodeAndEncryptProxyGrantingTicket(attributes);
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Initialize cipher based on service public key.
+     *
+     * @param service the service
+     * @return the false if no public key is found
+     * or if cipher cannot be initialized, etc.
+     */
+    protected boolean initializeCipherBasedOnServicePublicKey(final RegisteredService service) {
+        try {
+            final PublicKey publicKey = service.getPublicKey().createInstance();
+            if (publicKey == null) {
+                logger.debug("No public key is defined for service [{}]. No encoding will take place.", service);
+                return false;
+            }
+            logger.debug("Using public key [{}] to initialize the cipher", service.getPublicKey());
+
+            this.cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+            logger.debug("Initialized cipher in encrypt-mode via the public key algorithm [{}]",
+                    publicKey.getAlgorithm());
+            return true;
+        } catch (final Exception e) {
+            logger.warn(e.getMessage(), e);
+        }
+        return false;
+    }
+
+    @Override
+    public Map<String, Object> encodeAttributes(final Map<String, Object> attributes, final Service service) {
+        logger.debug("Starting to encode attributes for release to service [{}]", service);
+        final Map<String, Object> newEncodedAttributes = new HashMap<>(attributes);
+
+        initialize(newEncodedAttributes);
+
+        final RegisteredService registeredService = this.servicesManager.findServiceBy(service);
+        if (registeredService != null && registeredService.getAccessStrategy().isServiceAccessAllowed()) {
+            encodeAttributesInternal(newEncodedAttributes, registeredService);
+        }
+        logger.debug("[{}] Encoded attributes are available for release to [{}]",
+                newEncodedAttributes.size(), service);
+        return newEncodedAttributes;
+    }
+
+    /**
+     * Initialize the encoding process. Removes the
+     * {@link CasViewConstants#MODEL_ATTRIBUTE_NAME_PRINCIPAL_CREDENTIAL}
+     * and
+     * {@link CasViewConstants#MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET}
+     * from the authentication attributes originally and into a cache object, so it
+     * can later on be encrypted if needed.
+     * @param attributes the new encoded attributes
+     */
+    protected void initialize(final Map<String, Object> attributes) {
```

Comment:
I'm not sure to understand what this method does. password and pgt are moved to a specific collection for a future encryption, right?
In that case, how can it be thread-safe if the ```cachedAttributesToEncode``` is shared for the whole  ```DefaultCasAttributeEncoder```?